backend/accounts/transaction.go

```diff
@@ -133,5 +142,94 @@ type OrderedTransactions []*TransactionData
 // The input list is modified in place and must not be used anymore after calling this function.
 func NewOrderedTransactions(txs []*TransactionData) OrderedTransactions {
 	sort.Sort(sort.Reverse(byHeight(txs)))
+
+	balance := big.NewInt(0)
+	for i := len(txs) - 1; i >= 0; i-- {
+		tx := txs[i]
+		switch tx.Type {
+		case TxTypeReceive:
+			balance.Add(balance, tx.Amount.BigInt())
+		case TxTypeSend:
+			balance.Sub(balance, tx.Amount.BigInt())
+			// Subtract fee as well.
+			if tx.Fee != nil {
+				balance.Sub(balance, tx.Fee.BigInt())
+			}
+		case TxTypeSendSelf:
+			// Subtract only fee.
+			if tx.Fee != nil {
+				balance.Sub(balance, tx.Fee.BigInt())
+			}
+		}
+		tx.Balance = coin.NewAmount(balance)
+	}
 	return txs
 }
+
+// TimeseriesEntry contains the balance of the account at the given time.
+type TimeseriesEntry struct {
+	Time  time.Time
+	Value coin.Amount
+}
+
+// MarshalJSON serializes the entry as JSON.
+func (c *TimeseriesEntry) MarshalJSON() ([]byte, error) {
+	value, err := c.Value.Int64()
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(struct {
+		Time  int64 `json:"time"`
+		Value int64 `json:"value"`
+	}{
+		Time:  c.Time.Unix(),
+		Value: value,
+	})
+}
+
+// Timeseries chunks the time between `start` and `end` into steps of `interval` duration, and
+// provides the balance of the account at each step.
+func (txs OrderedTransactions) Timeseries(
+	start, end time.Time, interval time.Duration) ([]TimeseriesEntry, error) {
+	if len(txs) == 0 {
+		return nil, nil
+	}
+	for _, tx := range txs {
+		if tx.Height != 0 && tx.Timestamp == nil {
+			return nil, errp.WithStack(errors.ErrNotAvailable)
+		}
+	}
+	currentTime := start
+	if currentTime.IsZero() {
+		return nil, nil
+	}
+
+	result := []TimeseriesEntry{}
+	for {
+		// Find the latest tx before `currentTime`.
+		nextIndex := sort.Search(len(txs), func(idx int) bool {
+			tx := txs[idx]
+			if tx.Height == 0 {
+				return false
+			}
+			return txs[idx].Timestamp.Before(currentTime)
```

Comment:
or use the `tx` var since it's there anyway? `return tx.Timestamp.Before(currentTime)`

Reply:
oh, I think you'll actually want `return tx.Timestamp.Before(currenTime) || tx.Timestamp.Equal(currentTime)`. if there's a transaction which falls exactly on the `currentTime`, you'll want to pick its balance, not the one before, right.

But that condition is harder to read. A better equivalent would be: `return !tx.Timestamp.After(currentTime)` or `return !currentTime.Before(tx.Timestamp)`.

Reply:
actually, maybe `tx.Timestamp.Before(currenTime) || tx.Timestamp.Equal(currentTime)` is easer to read in the end.

Reply:
done